crates/bevy_ecs/src/system/query.rs

```diff
@@ -543,6 +543,18 @@ where
             >())),
         }
     }
+
+    /// Returns true if this query contains no elements.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
```

Comment:
This isn't safe. If some calls query.is_empty() in parallel on a mutable query, we have created aliased muts:

```rust
fn test(query: Query<&mut u32>, tasks: Res<ComputeTaskPool>) {
    tasks.scope(|s| {
        let q = &query;
        s.spawn(async move {
            q.is_empty();
        });
        s.spawn(async move {
            q.is_empty();
        })
    });
}
```

Reply:
The user doesn't have _access_ to them, but the muts exist in memory.

Reply:
@bjorn3 im assuming thats not good right :)

Reply:
haha, that's awkward 
Looks like I may have to implement this at the `QueryState` level then

Reply:
This might actually be a bit challenging. Its easy to implement this for unfiltered queries (just iterate matching archetypes and count the entities in each archetype). But filters require accessing per-entity data (in some cases), which might be mutated elsewhere (ex: Changed state).

The easy fix is to make `is_empty()` a mutable borrow, but that does restrict the contexts it can be used in. We could restrict `is_empty()` to "read-only queries" then add `is_empty_mut()` for mutable queries, but thats just ugly + confusing.

We could make is_empty() only evaluate filters (basically re-implement iteration logic, but without actually evaluating component fetches and short circuiting on the first filter match), which ensures no mutable references are created ... but thats complicated :)

No easy answers here.

Reply:
Can we just borrow from `query.single()` and try asking for the next entity? If it's Some, return false.

Reply:
Nope! That also creates a mutable reference :)

Reply:
This runs into the same problem though.
`single()` requires it to be a read only query.
so we'd have to have a `is_empty` that call `single` and an `is_empty_mut` that calls `single_mut`

Reply:
@cart I recently made a change where I tried to implement this at the `QueryIter` level, so hopefully this version never creates muts? (This is assuming I was able to understand the implementation all in the past hour or so lol)

Reply:
Also as a side note, I would like to point out that while reading through the ECS implementation, it was quite difficult to understand _especiall_ due to the lack of documentation at the lower level ECS functions.

Reply:
> Nope! That also creates a mutable reference :)

Darn. If we solve our inability to use `query.iter()` on mutable queries that should fix this as a side effect too presumably?

Reply:
Yup!

Reply:
> @cart I recently made a change where I tried to implement this at the QueryIter level, so hopefully this version never creates muts? (This is assuming I was able to understand the implementation all in the past hour or so lol)

Yup that's what I had in mind!